kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -208,6 +257,83 @@ def abstract_cell_vars(cterm: KInner, keep_vars: Collection[KVariable] = ()) ->
     return Subst(subst)(config)
 
 
+def check_implication(kcfg_explore: KCFGExplore, concrete: CTerm, abstract: CTerm) -> Tuple[bool, str]:
+    def _is_cell_subst(csubst: KInner) -> bool:
+        if type(csubst) is KApply and csubst.label.name == '_==K_':
+            csubst_arg = csubst.args[0]
+            if type(csubst_arg) is KVariable and csubst_arg.name.endswith('_CELL'):
+                return True
+        return False
+
+    def _is_negative_cell_subst(constraint: KInner) -> bool:
+        constraint_bool = ml_pred_to_bool(constraint)
+        if type(constraint_bool) is KApply and constraint_bool.label.name == 'notBool_':
+            negative_constraint = constraint_bool.args[0]
+            if type(negative_constraint) is KApply and negative_constraint.label.name == '_andBool_':
+                constraints = flatten_label('_andBool_', negative_constraint)
+                cell_constraints = list(filter(_is_cell_subst, constraints))
+                if len(cell_constraints) > 0:
+                    return True
+        return False
+
+    concrete_config, *concrete_constraints = concrete
+    abstract_config, *abstract_constraints = abstract
+    config_match = abstract_config.match(concrete_config)
+    if config_match is None:
+        _, concrete_subst = split_config_from(concrete_config)
+        cell_names = concrete_subst.keys()
+        failing_cells = []
+        for cell in cell_names:
+            concrete_cell = get_cell(concrete_config, cell)
+            abstract_cell = get_cell(abstract_config, cell)
```

Comment:
```suggestion
            concrete_cell = concrete.cell(cell)
            abstract_cell = abstract.cell(cell)
```

This variant where we use the `CTerm.cell` interace is automatically memoizing.

Reply:
This method was moved to pyk and this issue was addressed therre.